File: backends/go-gin/internal/infrastructure/cache/redis_invitation_auth_repository.go
package cache

import (
	"authorization-go/internal/config"
	"context"
	"strconv"
	"time"

	"github.com/redis/go-redis/v9"
)

// RedisInvitationAuthRepository は domain/invitation.AuthRepository の Redis 実装です。
type RedisInvitationAuthRepository struct {
	rdb *redis.Client
	cfg *config.Config
}

// NewRedisInvitationAuthRepository は RedisInvitationAuthRepository を生成します。
func NewRedisInvitationAuthRepository(rdb *redis.Client, cfg *config.Config) *RedisInvitationAuthRepository {
	return &RedisInvitationAuthRepository{rdb: rdb, cfg: cfg}
}

// Store はトークンとロールを指定秒数 Redis にキャッシュします。
func (r *RedisInvitationAuthRepository) Store(token string, role int, ttl int) error {
	ctx := context.Background()
	return r.rdb.Set(ctx, r.key(token), strconv.Itoa(role), time.Duration(ttl)*time.Second).Err()
}

// Find はキャッシュからロールを取得します。存在しない場合は nil を返します。
func (r *RedisInvitationAuthRepository) Find(token string) (*int, error) {
	ctx := context.Background()
	val, err := r.rdb.Get(ctx, r.key(token)).Result()
	if err == redis.Nil {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	role, err := strconv.Atoi(val)
	if err != nil {
		return nil, err
	}
	return &role, nil
}

// Remove はキャッシュからトークンを削除します。
func (r *RedisInvitationAuthRepository) Remove(token string) error {
	ctx := context.Background()
	return r.rdb.Del(ctx, r.key(token)).Err()
}

func (r *RedisInvitationAuthRepository) key(token string) string {
	prefix := r.cfg.App.CachePrefix
	if prefix == "" {
		return "invitation_auth:invitation_auth:" + token
	}
	return prefix + ":invitation_auth:invitation_auth:" + token
}

File: backends/ruby-hanami/app/infrastructure/cache/redis_invitation_auth_repository.rb
# frozen_string_literal: true
#
# Redis を用いた招待認証キャッシュリポジトリの実装モジュール。
#
# @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>

module Infrastructure
  module Cache
    # Redis を用いた招待認証キャッシュリポジトリの実装クラスです。
    # @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
    class RedisInvitationAuthRepository
      include Domain::Invitation::AuthRepository

      # @param cfg [ConfigLoader] 設定
      def initialize(cfg)
        @redis  = RedisClient.new_client(cfg)
        @prefix = cfg.app.cache_prefix
      end

      def store(token, role, ttl)
        @redis.set(cache_key(token), role.to_s, ex: ttl)
        nil
      end

      def find(token)
        value = @redis.get(cache_key(token))
        value&.match?(/\A\d+\z/) ? value.to_i : nil
      end

      def remove(token)
        @redis.del(cache_key(token))
        nil
      end

      private

      def cache_key(token)
        "#{@prefix}:invitation_auth:invitation_auth:#{token}"
      end
    end
  end
end
